### DISCUSSION TITLE
How to add additional metadata to users table?

### DISCUSSION THREAD

**Discussion by vedantroy** [supabase/supabase#6363] [Questions] [answered]
If I was implementing authentication myself, I would have a column in my users table called `onboarded`, or something like that, which would be set to `false` when a new users signs up.

Then, whenever a user visits a route in my web-app, I could check if they are on-boarded, and if not, redirect them to a tutorial.
What is the equivalent way of doing this in supabase, where the auth table is not created by me?

**Comment by GaryAustin1** [comment, marked as answer]
You could use the user meta data and update it server side. The recommendation though normally is to have your own copy of key user info from the auth.users table and your own user table.  This table is updated with a trigger function on auth.users inserts/updates/deletes.

**Reply by tanius** [reply]
I believe this should read 

> "The recommendation though normally is to have your own copy of key user info from the auth.users table **_in_** your own user table."

So not "… and your own user table".

**Comment by oalbacha** [comment]
Thanks for building and maintaining this awesome tool!

I have a public users table triggering on auth users signup sign-in. However, in case the user key information needs to be updated client-side, how will that affect the user's auth table. Is the recommendation to have a reverse trigger that changes the auth user metadata when the changes on public users happen client-side?

**Reply by silentworks** [reply]
No changes should be made to the `auth.users` table, there is already a foreign key in your `public.users` table that defines the relationship. The `auth.users` table should not be modified outside of using the supabase-js auth methods and most of the information in that table is sensitive data so it shouldn't be shared with your other table either.

**Reply by pnadj** [reply]
@silentworks @GaryAustin1  but won't data in the metadata be incorrect and outdated then? Do we just ignore it?

**Reply by silentworks** [reply]
@pnadj you can use `supabase.auth.updateUser` to update the metadata, I for one only use metadata as a temporary transport column. So I don't rely on data from there after the user has signed up.

**Comment by ghost** [comment]
In case anyone is trying to add metadata information from signUp({},{data:{...}}) [this](https://database.dev/handle-new-user) is how it can be handled.

**Reply by devidw** [reply]
In the docs: https://supabase.com/docs/reference/javascript/auth-signup#sign-up-with-additional-user-metadata

```ts
const { data, error } = await supabase.auth.signUp(
  {
    email: 'example@email.com',
    password: 'example-password',
    options: {
      data: {
        first_name: 'John',
        age: 27,
      }
    }
  }
)
```

As payload in a post request:

```
{
    "email": "",
    "password": "",
    "data": {
        "first_name": "John",
        "last_name": "Doe",
    }
}
```

**Reply by Sathish-Kumar-A** [reply]
but it is not working when I'm calling this from backend. magic link sending is happening but user metadata can't able to set

**Reply by devidw** [reply]
@Sathish-Kumar-A have a look into ~`auth.users.raw_user_metadata`~ `auth.users.raw_user_meta_data` column

**Reply by DRAHUL-30** [reply]
This is sathish kumar only, where to find this auth.users.raw_user_metdata in supabase

**Reply by devidw** [reply]
```sql
select raw_user_meta_data from auth.users
```

**Reply by Sathish-Kumar-A** [reply]
When I runs that query this error is coming
 error: {
    code: '42P01',
    details: null,
    hint: null,
    message: 'relation "public.auth.users" does not exist'
  }

and one more doubt, are you calling this function from backend(server side) or from client. 

const { email, password, role, phone } = req.body;
        const data = { role };
        const { user, session, error } = await Supabase.auth.signUp({
            email,
            password,
            options: {
                data:data
            }
        });

when I called this signup function in the backend, In the user variable , the user_metadata field is empty


this is response after signing in the user variable
{
  id: '7d87401e-6008-474e-aeba-1a6af07634e7',
  aud: 'authenticated',
  role: 'authenticated',
  email: 'sathish252199@gmail.com',
  phone: '',
  confirmation_sent_at: '2022-10-17T05:07:48.098085813Z',
  app_metadata: { provider: 'email', providers: [ 'email' ] },
  user_metadata: {}, //here I can't find any of my metadata here
  identities: [

**Reply by wizzymotion** [reply]
I'm facing same problem... any solution???

**Reply by devidw** [reply]
There is no _public.auth.users_ table, there only is an _auth.users_ table.

**Reply by treading0light** [reply]
> In case anyone is trying to add metadata information from signUp({},{data:{...}}) [this](https://database.dev/handle-new-user) is how it can be handled.

Thanks for the link. So with:  signUp({credentials}, {data: {name: 'tony'}), the name will be available in the Supabase function as: new.raw_user_meta_data.name ?

**Reply by rbrueckner** [reply]
I've been digging into the docs and github for hours now: where is the signature signUp({},{data:{...}}) documented?
Note that the second argument is separate dict {data:{...}}, which seems to be different from earlier docs.

**Reply by GaryAustin1** [reply]
@rbrueckner https://supabase.com/docs/reference/javascript/auth-signup 2nd example.

**Reply by rbrueckner** [reply]
@GaryAustin1: Thanks for the link, but I already knew it.
However, there it explicitely states:
const { data, error } = await supabase.auth.signUp(
  {
    email: 'example@email.com',
    password: 'example-password',
    options: {
      data: {
        first_name: 'John',
        age: 27,
      }
    }
  }
)
while @treading0light above mentioned that instead it needs to read
const { data, error } = await supabase.auth.signUp(
{
    {
        email: 'example@email.com',
        password: 'example-password',
    }
    {
        data: {
            first_name: 'John',
            age: 27,
        }
    }
)
Note the extra {} around the first block.
I've seen this as well in another post - that second seems to work, the first one doesn't.
And hence I wonder if the docs are outdated!?

**Reply by GaryAustin1** [reply]
@rbrueckner 
I think you might be looking at v1 versus v2.  v2 is what I intending to link (and get following the link).
v1 did not have the options reference.

This is what I use:
```
const result = await supabase1.auth.signUp({
         email: 'xxxx',
         password:'xxxx',
         options:{
            data:{
                hi:'test'
                }
         }
    })
```

**Reply by tanius** [reply]
If anyone needs a raw HTTP example, here is the format I had success with:

```
curl -X POST 'https://my-project-ref.supabase.co/auth/v1/signup' \
-H "apikey: SUPABASE_ANON_KEY" \
-H "Content-Type: application/json" \
-d '{
  "email": "test-1@example.com", 
  "password": "my-password-here",
  "data": { 
    "fullname": "Test User 1",
    "organization": "My Beautiful Org",
    "ui_language": "en"
  }
}'
```

Note how the JSON structure is different from what the JavaScript Supabase Client v1 and v2 expect. I did not find that documented, so here it goes. It follows the same JSON structure as documented for updating a user via `PUT /auth/v1/user` though.

**Comment by BirdyLeads** [comment]
how do you edit the public.users tables directly?  I know that there is a trigger and a function on signup to add user meta to the public.users table.  Now I want to add additional data to the the public.users table directly.  Is there a way to do that?
